p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,790 @@
+	// connections
+}
+
+type namespacedRoundTripper struct {
+	http.RoundTripper
+	protocolPrefix    string
+	protocolPrefixRaw string
+}
+
+func (rt *namespacedRoundTripper) GetPeerMetadata() (PeerMeta, error) {
+	if g, ok := rt.RoundTripper.(PeerMetadataGetter); ok {
+		return g.GetPeerMetadata()
+	}
+
+	return nil, fmt.Errorf("can not get peer protocol map. Inner roundtripper does not implement GetPeerMetadata")
+}
+
+// RoundTrip implements http.RoundTripper.
+func (rt *namespacedRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	if !strings.HasPrefix(r.URL.Path, rt.protocolPrefix) {
+		r.URL.Path = rt.protocolPrefix + r.URL.Path
+	}
+	if !strings.HasPrefix(r.URL.RawPath, rt.protocolPrefixRaw) {
+		r.URL.RawPath = rt.protocolPrefixRaw + r.URL.Path
+	}
+
+	return rt.RoundTripper.RoundTrip(r)
+}
+
+// NamespaceRoundTripper returns an http.RoundTripper that are scoped to the given protocol on the given server.
+func (h *Host) NamespaceRoundTripper(roundtripper http.RoundTripper, p protocol.ID, server peer.ID) (*namespacedRoundTripper, error) {
+	protos, err := h.getAndStorePeerMetadata(roundtripper, server)
+	if err != nil {
+		return &namespacedRoundTripper{}, err
+	}
+
+	v, ok := protos[p]
+	if !ok {
+		return &namespacedRoundTripper{}, fmt.Errorf("no protocol %s for server %s", p, server)
+	}
+
+	path := v.Path
+	if path[len(path)-1] == '/' {
+		// Trim the trailing slash, since it's common to make requests starting with a leading forward slash for the path
+		path = path[:len(path)-1]
+	}
+
+	u, err := url.Parse(path)
+	if err != nil {
+		return &namespacedRoundTripper{}, fmt.Errorf("invalid path %s for protocol %s for server %s", v.Path, p, server)
+	}
+
+	return &namespacedRoundTripper{
+		RoundTripper:      roundtripper,
+		protocolPrefix:    u.Path,
+		protocolPrefixRaw: u.RawPath,
+	}, nil
+}
+
+// NamespacedClient returns an http.Client that is scoped to the given protocol
+// on the given server. It creates a new RoundTripper for each call. If you are
+// creating many namespaced clients, consider creating a round tripper directly
+// and namespacing the roundripper yourself, then creating clients from the
+// namespace round tripper.
+func (h *Host) NamespacedClient(p protocol.ID, server peer.AddrInfo, opts ...RoundTripperOption) (http.Client, error) {
+	rt, err := h.NewConstrainedRoundTripper(server, opts...)
+	if err != nil {
+		return http.Client{}, err
+	}
+
+	nrt, err := h.NamespaceRoundTripper(rt, p, server.ID)
+	if err != nil {
+		return http.Client{}, err
+	}
+
+	return http.Client{Transport: nrt}, nil
+}
+
+// NewConstrainedRoundTripper returns an http.RoundTripper that can fulfill and HTTP
+// request to the given server. It may use an HTTP transport or a stream based
+// transport. It is valid to pass an empty server.ID.
+func (h *Host) NewConstrainedRoundTripper(server peer.AddrInfo, opts ...RoundTripperOption) (http.RoundTripper, error) {
```

Comment:
This function is a bit problematic, as it dials a new connection (which might take an extended period of time), without providing a way to cancel that. One option would be to add a context, but that would be a bit surprising for a function that look like a constructor. Another option would be to defer dialing.